### DISCUSSION TITLE
ZeroCopy both ways

### DISCUSSION THREAD

**Discussion by John0x** [fzyzcjy/flutter_rust_bridge#974] [General]
Hey, as far as I understand the documentation, ZeroCopy Buffers can only be used to send data from rust to dart and not the other way around, right?

Is there currently a way to send data from dart to rust, without copying? Let's say I have image data, managed by Dart, which I want to process using Rust. What would be the best approach?

**Comment by fzyzcjy** [comment]
That's a limitation of Dart VM API.

> Let's say I have image data, managed by Dart, which I want to process using Rust. What would be the best approach?

Workaround: Maybe let Rust *own* the image data, instead of Dart? For example, the "pool" or "opaque types" approach can allow you to let Rust own the data. However, you may need to check safety-related things.

**Reply by John0x** [reply]
That would be an option. Is there more documentation about the opaque types and the pool?
I only found some issues and abandoned pull requests, which is kinda confusing, to be honest. As far as I understand, the opaque types allow you to expose complex rust structs to Dart right? But they are passed by copy and and therefore not like the ZeroCopy buffers?

**Reply by fzyzcjy** [reply]
https://cjycode.com/flutter_rust_bridge/feature/lang_rust_opaque.html
https://cjycode.com/flutter_rust_bridge/feature/lang_dart_opaque.html
https://cjycode.com/flutter_rust_bridge/feature/object_pool.html

**Reply by fzyzcjy** [reply]
as for "passed by copy" I have updated doc just now

**Reply by John0x** [reply]
So, if I pass on opaque types as an argument to a rust function, the rust function will basically get a pointer to the memory managed by Dart?
And if I return a opaque types from Rust, the Dart side will receive a pointer to the memory managed by Rust?

Therefore, I could just (in theory) have a global array of the images in my Rust code and return them (when needed on the dart side) as an opaque type and the data wouldn't be copied?

Which would also mean, that I can take an opaque type as an argument in Rust and return that (after modifying it) to dart and would have zero copies of the data?

**Reply by fzyzcjy** [reply]
> So, if I pass on opaque types as an argument to a rust function, the rust function will basically get a pointer to the memory managed by Dart?
And if I return a opaque types from Rust, the Dart side will receive a pointer to the memory managed by Rust?

Yes

> Therefore, I could just (in theory) have a global array of the images in my Rust code and return them (when needed on the dart side) as an opaque type and the data wouldn't be copied?

No need for global array. That's for the "pool" scenario (if you like totally ok).

Just, say, return a `OpaqueRust<YourBigImageType>`. Here you moved the big image (in the rust "move" sense) from your function to the flutter_rust_bridge system. No worries, the frb system will handle its lifecycle and not copy it and free it when suitable.

> Which would also mean, that I can take an opaque type as an argument in Rust and return that (after modifying it) to dart and would have zero copies of the data?

sure

**Reply by fzyzcjy** [reply]
However, you may need some trick (feel free to PR!) to get the *underlying* data from the opaque. For example, get the data inside a Dart opaque from Rust code, or vise versa.

**Reply by fzyzcjy** [reply]
Also cc @rogurotus who implemented that great opaque feature

**Reply by John0x** [reply]
Ah, okay. So the opaque types are "just" pointers? They don't expose any of the underlying data.
Given the code:
```rust
pub struct HelloStruct {
    pub foo: String,
}

pub fn return_some_struct() -> RustOpaque<HelloStruct> {
    todo!()
}
```

I would be able to call return_some_struct from dart and would get a pointer to the struct, but without a way of accessing the structs fields (like foo). So we should, theoretically, be able to read the memory at the address of that pointer to get access to the data of the struct ? It's my first time using Dart, so I'm not sure about how to approach this. But I'll look into it. Thank you :)

**Reply by fzyzcjy** [reply]
> They don't expose any of the underlying data.

Yes (currently), but you may use some hacks to get it.

> So we should, theoretically, be able to read the memory at the address of that pointer to get access to the data of the struct ?

I think so. Except that it is not a raw pointer but a wrapped pointer, so should shift a few bytes. Or use another rust method like `pub fn extract_dangerous_pointer_from_opaque(opaque: RustOpaque<HelloStruct>) -> *void { &opaque.inner.foo }` etc

I have to say that, this should be thoroughly considered before implementing (but surely trust your ability!), otherwise the lifetime can be a problem and you will get dangling ptr etc

You are welcome!
